File: sort/cpp/sort.cpp
#include "sort.hpp"
#include <algorithm>
template <class T>
bool SelectionSort<T>::sort(std::vector<T>& vec, int32_t low, int32_t high) {
    int32_t min = 0;
    for(int32_t i = 0; i < high; i++) {
        min = i;
        for(int32_t j = i + 1; j < high; j++) {
            if(vec[j] < vec[min]) { min = j; }
        }
        Sort<T>::swap(vec, i, min);
    }
    return false;
}

template <class T>
bool InsertionSort<T>::sort(std::vector<T>& vec, int32_t low, int32_t high) {
    int32_t j = 0;
    int32_t key = 0;
    for(int32_t i = 1; i < high; i++) {
        j = i - 1;
        key = vec[i];
        while(j >=0 && vec[j] > key) {
            vec[j + 1] = vec[j];
            j -= 1;
        }
        vec[j + 1] = key;
    }
    return false;
}

template <class T>
bool MergeSort<T>::sort(std::vector<T>& vec, int32_t low, int32_t high) {
    if(high <= low) { return false; }

    const int32_t mid = low + (high - low) / 2;
    sort(vec, low, mid);
    sort(vec, mid + 1, high);

    merge(vec, low, mid, high);
    return false;
}

template <class T>
bool QuickSort<T>::sort(std::vector<T>& vec, int32_t low, int32_t high) {
    if(high <= low) { return false; }
    
    const int32_t part = partition(vec, low, high);
    this->sort(vec, low, part - 1);
    this->sort(vec, part + 1, high);
    return false;
}
/**
 * Merge
 **/
template <class T>
void MergeSort<T>::merge(std::vector<T>& vec, const int32_t low, const int32_t mid, const int32_t high) {
    const int32_t leftHalfSize = mid - low + 1;
    const int32_t rightHalfSize = high - mid;

    std::vector<T> leftVec;
    std::vector<T> rightVec;

    int32_t i = 0;
    int32_t j = 0;
    int32_t k = low;

    for (i = 0; i < leftHalfSize; i++) 
        leftVec.emplace_back(vec[low + i]); 
    for (j = 0; j < rightHalfSize; j++) 
        rightVec.emplace_back(vec[mid + 1 + j]);

    i = j = 0;

    while(i < leftHalfSize && j < rightHalfSize) {
        if(leftVec[i] <= rightVec[j]) {
            vec[k] = leftVec[i];
            i++;
        } else {
            vec[k] = rightVec[j];
            j++;
        }
        k++;
    }

    while(i < leftHalfSize) {
        vec[k] = leftVec[i];
        i++;
        k++;
    }

    while(j < rightHalfSize) {
        vec[k] = rightVec[j];
        j++;
        k++;
    }
}

template <class T>
int32_t QuickSort<T>::partition(std::vector<T>& vec, const int32_t low, const int32_t high) {
    uint32_t i = low;
    uint32_t j = high + 1;
    const T tmp = vec[low];
    while(true) {
        while(vec[++i] < tmp) { 
            if(i == high) { break; }
        }
        while(tmp < vec[--j]) {
            if(j == low) { break; }
        }
        if(i >= j) { break; }
        Sort<T>::swap(vec, i, j);
    }
    Sort<T>::swap(vec, low, j);
    return j;
}

//
template class SelectionSort<int32_t>;
template class SelectionSort<int64_t>;
template class SelectionSort<float>;
//
template class InsertionSort<int32_t>;
template class InsertionSort<int64_t>;
template class InsertionSort<float>;
//
template class MergeSort<int32_t>;
template class MergeSort<int64_t>;
template class MergeSort<float>;
//
template class QuickSort<int32_t>;
template class QuickSort<int64_t>;
template class QuickSort<float>;

File: sort/js/sort.js
"use strict";

class SelectionSort {
    sort(array, low, high) {
        const N = high - low;
        let min = 0;
        for(let i = 0; i < N; i++) {
            min = i;
            for(let j = i + 1; j < N; j++) {
                if(array[j] < array[min]) { 
                    min = j;
                }
            }
            [array[i], array[min]] = [array[min], array[i]];
        }
    }
}

class InsertionSort {
    sort(array, low, high) {
        const N = high - low;
        let j = 0;
        let key = 0;
        for(let i = 0; i < N; i++) {
            j = i - 1;
            key = array[i];
            while(j >= 0 && array[j] > key) {
                array[j + 1] = array[j];
                j--;
            }
            array[j + 1] = key;
        }
    }
}

class MergeSort {
    sort(array, low, high) {
        if(high <= low) {
            return;
        }
        const mid = Math.floor(low + (high - low) / 2);
        this.sort(array, low, mid);
        this.sort(array, mid + 1, high);

        this.merge(array, low, mid, high);
    }

    merge(array, low, mid, high) {
        const leftHalfSize = mid - low + 1;
        const rightHalfSize = high - mid;

        let i = 0;
        let j = 0;
        let k = low;

        let left = new Array(leftHalfSize);
        let right = new Array(rightHalfSize);

        for (i = 0; i < leftHalfSize; i++) {
            left[i] = array[low + i]; 
        }

        for (j = 0; j < rightHalfSize; j++) {
            right[j] = array[mid + 1 + j];
        }

        i = j = 0;

        while(i < leftHalfSize && j < rightHalfSize) {
            if(left[i] <= right[j]) {
                array[k] = left[i];
                i++;
            } else {
                array[k] = right[j];
                j++;
            }
            k++;
        }

        while(i < leftHalfSize) {
            array[k] = left[i];
            i++;
            k++;
        }
    
        while(j < rightHalfSize) {
            array[k] = right[j];
            j++;
            k++;
        }
    }
}

class QuickSort {
    sort(array, low, high) {
        if(high <= low) {
            return
        }
        const part = this.partition(array, low, high)
        this.sort(array, low, part);
        this.sort(array, part + 1, high);
    }

    partition(array, low, high) {
        let i = low;
        let j = high + 1;
        const tmp = array[low];
        while(true) {
            while(array[++i] < tmp) { 
                if(i == high) { break; }
            }
            while(tmp < array[--j]) {
                if(j == low) { break; }
            }
            if(i >= j) { break; }
            [array[i], array[j]] = [array[j], array[i]];
        }
        [array[low], array[j]] = [array[j], array[low]];
        return j;
    }
}

function testSort(sort, name, decLen) {
    console.log("-=| TESTING " + name + " SORT |=-");
    let data = genRandomData(20);
    printArray(data);
    if(decLen) {
        sort.sort(data, 0, data.length - 1);
    } else {
        sort.sort(data, 0, data.length);
    }
    printArray(data);
}

function printArray(array) {
    console.log(array)
}

function getRandomInt(min, max) {
    min = Math.ceil(min);
    max = Math.floor(max);
    return Math.floor(Math.random() * (max - min)) + min;
}

function genRandomData(size) {
    let array = new Array(size);
    for(let i = 0; i < size; i++) {
        array[i] = getRandomInt(0, 1000);
    }
    return array;
}

let main = function() {
    let selectionSort = new SelectionSort();
    testSort(selectionSort, "SELECTION", false);
    let insSort = new InsertionSort();
    testSort(insSort, "INSERTION", false);
    let mergeSort = new MergeSort();
    testSort(mergeSort, "MERGE", true);
    let quickSort = new QuickSort();
    testSort(quickSort, "QUICK", true);
}

if (typeof require !== 'undefined' && require.main === module) {
    main();
}